metadata/src/main/java/org/apache/kafka/controller/QuorumController.java

```diff
@@ -1142,9 +1190,20 @@ public ControllerResult<Void> generateRecordsAndResult() throws Exception {
                         "at metadata.version {} from {}.", bootstrapMetadata.records().size(),
                         bootstrapMetadata.metadataVersion(), bootstrapMetadata.source());
                 records.addAll(bootstrapMetadata.records());
-            } else if (featureControl.metadataVersion().equals(MetadataVersion.MINIMUM_KRAFT_VERSION)) {
-                log.info("No metadata.version feature level record was found in the log. " +
+
```

Comment:
I was thinking that we should extract `CompleteActivationEvent#generateRecordsAndResult` into a static function that can be unit tested.

Maybe something like 
```
static List<ApiMessageAndVersion> generateActivationRecords(
  Logger log,
  boolean logIsEmpty,
  boolean zkMigrationEnabled,
  BootstrapMetadata bootstrapMetadata
) {
...
}
```

Then we could have some nice unit tests. As it is, it's getting too complex without them.